File: cs/ch16/Money/Sum.cs
using System.Net;

namespace TheSoftwareGorilla.TDD.Money;

public class Sum : Expression
{
    public Expression Augend { get; }
    public Expression Addend { get; }

    public Sum(Expression augend, Expression addend)
    {
        Augend = augend;
        Addend = addend;
    }

    public Money Reduce(Bank bank, string to)
    {
        int amount = Augend.Reduce(bank, to).Amount + Addend.Reduce(bank, to).Amount;
        return new Money(amount, to);
    }

    public Expression Plus(Expression addend)
    {
        return new Sum(this, addend);
    }

    public Expression Times(int multiplier)
    {
        return new Sum(Augend.Times(multiplier), Addend.Times(multiplier));
    }
}

File: java/ch15/src/main/java/com/thesoftwaregorilla/tdd/money/Sum.java
package com.thesoftwaregorilla.tdd.money;

public class Sum implements Expression {
    final Expression augend;
    final Expression addend;

    public Sum(Expression augend, Expression addend) {
        this.augend = augend;
        this.addend = addend;
    }

    public Money reduce(Bank bank, String to) {
        return new Money(augend.reduce(bank, to).getAmount() + addend.reduce(bank, to).getAmount(), to);
    }

    public Expression plus(Expression addend) {
        return null;
    }
}
